homeassistant/components/zha/core/helpers.py

```diff
@@ -136,3 +139,36 @@ def convert_ieee(ieee_str):
     """Convert given ieee string to EUI64."""
     from zigpy.types import EUI64, uint8_t
     return EUI64([uint8_t(p, base=16) for p in ieee_str.split(':')])
+
+
+def construct_unique_id(cluster):
+    """Construct a unique id from a cluster."""
+    ieee = cluster.endpoint.device.ieee
+    ieeetail = ''.join(['%02x' % (o, ) for o in ieee[-4:]])
+    manufacturer = cluster.endpoint.manufacturer
+    model = cluster.endpoint.model
+    unique_id = None
+    if manufacturer and model is not None:
+        unique_id = "{}_{}_{}_{}_{}".format(
+            slugify(manufacturer),
+            slugify(model),
+            ieeetail,
+            cluster.endpoint.endpoint_id,
+            cluster.cluster_id,
+        )
+    else:
+        unique_id = "zha_{}_{}_{}".format(
+            ieeetail,
+            cluster.endpoint.endpoint_id,
+            cluster.cluster_id,
+        )
+    return unique_id
+
+
+def get_attr_id_by_name(cluster, attr_name):
+    """Get the attribute id for a cluster attribute by its name."""
+    cluster_attributes = {}
+    # pylint: disable=W0612
+    for attrid, (attrname, datatype) in cluster.attributes.items():
+        cluster_attributes[attrname] = attrid
```

Comment:
Why all these assignments if you only care about the value for `attrname` ?

Reply:
It’s a reverse lookup off of a structure from Zigpy. I have to unwind that structure so I can perform the lookup by name instead of by Id.

Reply:
```
def get_attr_id_by_name(cluster, attr_name):
    """Get the attribute id for a cluster attribute by its name."""
    # pylint: disable=W0612
    for attrid, (attrname, datatype) in cluster.attributes.items():
        if attr_name == attrname:
            return attrid
    return None
```

better?

Reply:
or this?
```
def get_attr_id_by_name(cluster, attr_name):
    """Get the attribute id for a cluster attribute by its name."""
    return next((attrid for attrid, (attrname, datatype) in
                 cluster.attributes.items() if attr_name == attrname), None)
```

I like this one better myself but I know some don't like the syntax.

Reply:
BTW, Zigpy does maintain `cluster._attridx['attr_name']` dict with attr name to attr_id mapping. Maybe it is time to expose it through a public property.

Reply:
It does but I can’t use it without a release... I’ll swap to it as soon as it’s available